src/App.jsx

```diff
@@ -14,45 +17,105 @@ export const goodsFromServer = [
-        className="button is-success is-light"
-      >
-        Sort by length
-      </button>
-
-      <button
-        type="button"
-        className="button is-warning is-light"
-      >
-        Reverse
-      </button>
-
-      <button
-        type="button"
-        className="button is-danger is-light"
-      >
-        Reset
-      </button>
-    </div>
+const SORT_FIELD_LENGTH = 'length';
+const SORT_FIELD_ALPHABETICALLY = 'alphabetically';
+
+const getPreparedGoods = (goods, { sortParam, isReversed }) => {
+  const prepareGoods = [...goods];
+
+  if (sortParam) {
+    prepareGoods.sort((good1, good2) => {
+      switch (sortParam) {
+        case SORT_FIELD_ALPHABETICALLY:
+          return good1.localeCompare(good2);
+
+        case SORT_FIELD_LENGTH:
+          return good1.length - good2.length;
+
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (isReversed) {
+    prepareGoods.reverse();
+  }
+
+  return prepareGoods;
+};
 
-    <ul>
-      <li data-cy="Good">Dumplings</li>
-      <li data-cy="Good">Carrot</li>
-      <li data-cy="Good">Eggs</li>
-      <li data-cy="Good">Ice cream</li>
-      <li data-cy="Good">Apple</li>
-      <li data-cy="Good">...</li>
-    </ul>
-  </div>
-);
+export const App = () => {
+  const [sortParam, setSortParam] = useState(null);
+  const [isReversed, setIsReversed] = useState(false);
+  const reparedGoods = getPreparedGoods(
+    goodsFromServer,
+    { sortParam, isReversed },
+  );
+
+  return (
+    <div className="section content">
+      <div className="buttons">
+        <button
+          type="button"
+          className={classes(
+            'button',
+            'is-info',
+            { 'is-light': sortParam !== SORT_FIELD_ALPHABETICALLY },
+          )}
+          onClick={() => setSortParam(SORT_FIELD_ALPHABETICALLY)}
+        >
+          Sort alphabetically
+        </button>
+
+        <button
+          type="button"
+          className={classes(
+            'button',
+            'is-success',
+            { 'is-light': sortParam !== SORT_FIELD_LENGTH },
+          )}
+          onClick={() => setSortParam(SORT_FIELD_LENGTH)}
+        >
+          Sort by length
+        </button>
+
+        <button
+          type="button"
+          className={classes(
+            'button',
+            'is-warning',
+            { 'is-light': !isReversed },
+          )}
+          onClick={() => setIsReversed(!isReversed)}
+        >
+          Reverse
+        </button>
+
+        {(sortParam || isReversed)
+          && (
+            <button
+              type="button"
+              className="button is-danger is-light"
+              onClick={() => {
+                setIsReversed(false);
+                setSortParam(null);
+              }
+              }
+            >
+              Reset
+            </button>
+          )
+        }
+      </div>
+
+      <ul>
+        {reparedGoods.map(good => (
+          <li data-cy="Good" key={uuidv4()}>
```

Comment:
React uses keys to efficiently update and re-render components in lists. Keys should be stable across re-renders to help React identify which items have changed, been added, or been removed. If you generate new keys every time you render, React may treat all items as new

Reply:
Thanks for the really cool advice! I know how and for what is `key' but I can't explain why I used it this way ), now I realize my mistake.